File: ruby/spec/02_linked_lists/2_1_remove_dups.rb
# n - list size
# time = O(n), mem = O(n)
def remove_dups_b(list)
  s = Set.new
  curr = list

  while curr do
    s.add(curr.value)
    curr = curr.next_node
  end

  curr = list
  prev = nil

  while curr do
    if s.include?(curr.value)
      s.delete(curr.value)
      prev = curr
    else
      prev.next_node = curr.next_node
    end

    curr = curr.next_node
  end

  list
end

# time = O(n*log(n)), mem = O(n)
def remove_dups_c(list)
  head = sort_ll(list)
  curr = head

  while curr.next_node do
    if curr.value == curr.next_node.value
      curr.next_node = curr.next_node.next_node
    else
      curr = curr.next_node
    end
  end

  head
end

def sort_ll(head)
  if ll_size(head) < 2
    head
  else
    a, b = *halve_list(head)
    merge_sorted_ll(sort_ll(a), sort_ll(b))
  end
end

def halve_list(head)
  half = ll_size(head) / 2
  a = head
  curr = a
  a_size = 1

  while a_size < half
    curr = curr.next_node
    a_size += 1
  end

  b = curr.next_node
  curr.next_node = nil
  [a, b]
end

def merge_sorted_ll(a, b)
  head = nil

  if a.value < b.value
    head = a
    a = a.next_node
  else
    head = b
    b = b.next_node
  end

  curr = head

  while a || b do
    if a && b
      if a.value < b.value
        curr.next_node = a
        a = a.next_node
      else
        curr.next_node = b
        b = b.next_node
      end
    elsif a
      curr.next_node = a
      a = a.next_node
    else
      curr.next_node = b
      b = b.next_node
    end

    curr = curr.next_node
  end

  head
end

RSpec.describe 'remove_dups' do
  %i(b c).each do |implementation|
    describe "#{implementation} case" do
      subject { send("remove_dups_#{implementation}", list) }

      context 'couple of dups' do
        let(:list) { ll([1, 2, 3, 1, 3, 4]) }

        it { expect(ll_to_array(subject)).to eq [1, 2, 3, 4] }
      end

      context 'many dups' do
        let(:list) { ll([1, 1, 3, 1, 3, 4, 3]) }

        it { expect(ll_to_array(subject)).to eq [1, 3, 4] }
      end
    end
  end
end

File: go/02_linked_lists/2_1_remove_dups.go
package linked_lists

import (
	"github.com/golang-collections/collections/set"
)

// n - list size
// time = O(n), mem = O(n)
func RemoveDupsB(head *llnode) *llnode {
	s := set.New()
	curr := head

	for curr != nil {
		s.Insert(curr.value)
		curr = curr.next
	}

	curr = head
	var prev *llnode

	for curr != nil {
		if s.Has(curr.value) {
			s.Remove(curr.value)
			prev = curr
		} else {
			prev.next = curr.next
		}

		curr = curr.next
	}

	return head
}

// time = O(n*log(n)), mem = O(n)
func RemoveDupsC(head *llnode) *llnode {
	head = SortLL(head)
	curr := head

	for curr.next != nil {
		if curr.value == curr.next.value {
			curr.next = curr.next.next
		} else {
			curr = curr.next
		}
	}

	return head
}

func SortLL(head *llnode) *llnode {
	if head.LLSize() < 2 {
		return head
	} else {
		a, b := HalveList(head)
		return MergeSortedLL(SortLL(a), SortLL(b))
	}
}

func HalveList(head *llnode) (*llnode, *llnode) {
	half := head.LLSize() / 2
	a := head
	curr := a
	aSize := 1

	for aSize < half {
		curr = curr.next
		aSize += 1
	}

	b := curr.next
	curr.next = nil
	return a, b
}

func MergeSortedLL(a *llnode, b *llnode) *llnode {
	var head *llnode

	if a.value < b.value {
		head = a
		a = a.next
	} else {
		head = b
		b = b.next
	}

	curr := head

	for a != nil || b != nil {
		if a != nil && b != nil {
			if a.value < b.value {
				curr.next = a
				a = a.next
			} else {
				curr.next = b
				b = b.next
			}
		} else if a != nil {
			curr.next = a
			a = a.next
		} else {
			curr.next = b
			b = b.next
		}

		curr = curr.next
	}

	return head
}
